glib/src/object.rs

```diff
@@ -3376,6 +3401,65 @@ wrapper! {
     }
 }
 
+type WeakRefNotifyData = ManuallyDrop<Pin<Box<Option<Box<dyn FnOnce() + 'static>>>>>;
+
+// rustdoc-stripper-ignore-next
+/// A handle to disconnect a weak ref notify closure.
+pub struct WeakRefNotify<T: ObjectType> {
+    object: WeakRef<T>,
+    data: WeakRefNotifyData,
+}
+
+unsafe extern "C" fn notify_func(data: ffi::gpointer, _obj: *mut gobject_ffi::GObject) {
+    let callback: Box<Option<Box<dyn FnOnce()>>> = Box::from_raw(data as *mut _);
+
+    let callback = (*callback).expect("cannot get closure...");
+    callback()
+}
+
+impl<T: ObjectType> WeakRefNotify<T> {
+    fn new<F: FnOnce() + 'static>(obj: &T, f: F) -> WeakRefNotify<T> {
+        unsafe {
+            let data: WeakRefNotifyData = ManuallyDrop::new(Box::pin(Some(Box::new(f))));
+
+            let data_ptr: *const Option<Box<dyn FnOnce()>> = Pin::as_ref(&data).get_ref();
+            gobject_ffi::g_object_weak_ref(
+                obj.as_ptr() as *mut GObject,
+                Some(notify_func),
+                data_ptr as *mut _,
+            );
+
+            WeakRefNotify {
+                object: obj.unsafe_cast_ref::<T>().downgrade(),
+                data,
+            }
+        }
+    }
+
+    // rustdoc-stripper-ignore-next
+    /// Try to upgrade this weak reference to a strong reference.
+    ///
+    /// If the stored object was already destroyed then `None` is returned.
+    pub fn upgrade(&self) -> Option<T> {
+        self.object.upgrade()
+    }
+
+    #[doc(alias = "g_object_weak_unref")]
+    pub fn disconnect(mut self) {
+        unsafe {
+            if let Some(obj) = self.object.upgrade() {
+                let data_ptr: *const Option<Box<dyn FnOnce()>> = Pin::as_ref(&self.data).get_ref();
+                gobject_ffi::g_object_weak_unref(
+                    obj.as_ptr() as *mut GObject,
+                    Some(notify_func),
+                    data_ptr as *mut _,
+                );
+                ManuallyDrop::drop(&mut self.data);
```

Comment:
And also here an explanation why it's safe to drop the data (i.e. why it's guaranteed to be still alive, there are no race conditions, etc).